registry/storage/driver/middleware/bunnycdn/middleware.go

```diff
@@ -0,0 +1,118 @@
+package bunnycdn
+
+import (
+	"context"
+	"crypto/md5"
+	"encoding/base64"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	dcontext "github.com/docker/distribution/context"
+	storagedriver "github.com/docker/distribution/registry/storage/driver"
+	storagemiddleware "github.com/docker/distribution/registry/storage/driver/middleware"
+)
+
+type bunnyCDNStorageMiddleware struct {
+	storagedriver.StorageDriver
+	baseURL  string
+	token    string
+	duration time.Duration
+}
+
+var _ storagedriver.StorageDriver = &bunnyCDNStorageMiddleware{}
+
+// newBunnyCDNStorageMiddleware constructs and returns a new BunnyCDN
+// StorageDriver implementation.
+// Required options: baseurl, token
+// Optional options: duration
+func newBunnyCDNStorageMiddleware(storageDriver storagedriver.StorageDriver, options map[string]interface{}) (storagedriver.StorageDriver, error) {
+	// parse baseurl
+	base, ok := options["baseurl"]
+	if !ok {
+		return nil, fmt.Errorf("no baseurl provided")
+	}
+	baseURL, ok := base.(string)
+	if !ok {
+		return nil, fmt.Errorf("baseurl must be a string")
+	}
+	if !strings.Contains(baseURL, "://") {
+		baseURL = "https://" + baseURL
+	}
+	if _, err := url.Parse(baseURL); err != nil {
+		return nil, fmt.Errorf("invalid baseurl: %v", err)
+	}
+
+	baseURL = strings.TrimSuffix(baseURL, "/")
+
+	// parse token
+	at, ok := options["token"]
+	if !ok {
+		return nil, fmt.Errorf("no authtype provided")
+	}
+	token, ok := at.(string)
+	if !ok {
+		return nil, fmt.Errorf("token must be a string")
+	}
+
+	// parse duration
+	duration := 20 * time.Minute
+	d, ok := options["duration"]
+	if ok {
+		switch d := d.(type) {
+		case time.Duration:
+			duration = d
+		case string:
+			dur, err := time.ParseDuration(d)
+			if err != nil {
+				return nil, fmt.Errorf("invalid duration: %s", err)
+			}
+			duration = dur
+		}
+	}
+
+	return &bunnyCDNStorageMiddleware{
+		StorageDriver: storageDriver,
+		baseURL:       baseURL,
+		token:         token,
+		duration:      duration,
+	}, nil
+}
+
+// URLFor attempts to find a url which may be used to retrieve the file at the given path.
+func (bcm *bunnyCDNStorageMiddleware) URLFor(ctx context.Context, path string, options map[string]interface{}) (string, error) {
+
+	if bcm.StorageDriver.Name() != "s3aws" {
```

Comment:
nit: instead of hard-coding the `s3aws` driver name here we should probably make the [existing package level constant](https://github.com/docker/distribution/blob/2800ab02245e2eafc10e338939511dd1aeb5e135/registry/storage/driver/s3-aws/s3.go#L45) public and use it here.

Reply:
Thanks, this has been updated.